File: LeetCode/C++_Solutions/7_Reverse-Integer.cpp
#include <iostream>
using namespace std;

class Solution {
public:
    int reverse(int n) {
        int revNum = 0;
        while(n != 0){
            int extract = n % 10;
            if(revNum > INT32_MAX  / 10 || revNum == INT32_MAX  / 10 && extract > 7) return 0; // overflow check
            if(revNum < INT32_MIN  / 10 || revNum == INT32_MIN  / 10 && extract < -8) return 0; // underflow check
            revNum = (revNum * 10) + extract;
            n /= 10;
        }
        return revNum;
    }
};

File: LeetCode/Java_Solutions/7_Reverse-Integer.java
class Solution {
    public int reverse(int x) {
        int revNum = 0;
        while(x != 0){
            int extract = x % 10;
            if(revNum > Integer.MAX_VALUE / 10 || (revNum == Integer.MAX_VALUE / 10 && extract > 7)) return 0;
            if(revNum < Integer.MIN_VALUE / 10 || (revNum == Integer.MIN_VALUE / 10 && extract < -8)) return 0;
            revNum = (revNum * 10) + extract;
            x /= 10;
        }
        return revNum;
    }
}